### PR TITLE
Requester can mark a request as completed

### PR DESCRIPTION
**Pull request by kermit-klein**
[PT LINK](https://www.pivotaltracker.com/story/show/173333221)

- Add api/my_requests/requests/:id route to control user specific request update action
- Add status column to request, enums pending,active,completed
- Add happy and sad path tests
- Update request model test
- Create controller for my_requests with update action

Work done collaboration with Steve

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,10 +1,18 @@
 # frozen_string_literal: true
 
 Rails.application.routes.draw do
+  namespace :api do
+    namespace :my_request do
+      get 'requests/update'
+    end
+  end
```
Comment (erikbjoern) [inline config/routes.rb]: this is unnecessary

**Review by erikbjoern — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 1 above)
Comment (kermit-klein) [reply config/routes.rb]: yes autogenerated,
removed now

**Review by designerofthing — APPROVED** [review]

**Review by PaulineBA — APPROVED** [review]

**Review by erikbjoern — APPROVED** [review]

File: app/controllers/api/requests_controller.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # frozen_string_literal: true
 
 class Api::RequestsController < ApplicationController
-  before_action :authenticate_user!, only: [:create]
+  before_action :authenticate_user!, only: %i[create update]
```
Comment (ViaMarcus) [inline app/controllers/api/requests_controller.rb]: Did you forget to remove this when you realized you were doing it in another file?

File: app/models/request.rb [hunk 3]
```diff
@@ -1,7 +1,8 @@
 # frozen_string_literal: true
 
 class Request < ApplicationRecord
-  validates_presence_of :title, :description, :reward
+  validates_presence_of :title, :description, :reward, :status
   belongs_to :requester, class_name: 'User'
   has_many :offers
+  enum status: %i[pending active completed]
```
Comment (ViaMarcus) [inline app/models/request.rb]: Could you make the enum a hash, and save space in the numbers for "started"?

File: spec/requests/api/requests/user_can_mark_a_request_complete_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    it 'changes the status of the request' do
+      expect(request.status).to eq 'completed'
+    end
+
+    it 'responds with the completed confirmation' do
+      expect(response_json['message']).to eq 'Request completed!'
+    end
+  end
+
+  describe "User can't mark another user's request as completed" do
```
Comment (ViaMarcus) [inline spec/requests/api/requests/user_can_mark_a_request_complete_spec.rb:35]: "User can't mark a pending request as completed" would probably be a good sad path test

File: spec/requests/my_request_request_spec.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "MyRequests", type: :request do
```
Comment (ViaMarcus) [inline spec/requests/my_request_request_spec.rb]: Remove this if you are not going to use it

**Review by ViaMarcus — CHANGES_REQUESTED** [review]

File: app/models/request.rb (hunk 3 above)
Comment (kermit-klein) [reply app/models/request.rb]: Isn't pending request that a request waiting for a helper ?

File: spec/requests/api/requests/user_can_mark_a_request_complete_spec.rb (hunk 4 above)
Comment (erikbjoern) [reply spec/requests/api/requests/user_can_mark_a_request_complete_spec.rb:35]: Yes, Marcus's suggestion for a sad path is very relevant.

**Review by jysmys — APPROVED** [review]

File: app/models/request.rb (hunk 3 above)
Comment (ViaMarcus) [reply app/models/request.rb]: I meant "started" as in requester presses start in order to lock the job and make the request undeletable. But maybe we can just shove it in when we get to it.

File: spec/requests/api/requests/user_can_mark_a_request_complete_spec.rb (hunk 4 above)
Comment (kermit-klein) [reply spec/requests/api/requests/user_can_mark_a_request_complete_spec.rb:35]: done

File: app/controllers/api/requests_controller.rb (hunk 2 above)
Comment (kermit-klein) [reply app/controllers/api/requests_controller.rb]: done

File: spec/requests/my_request_request_spec.rb (hunk 5 above)
Comment (kermit-klein) [reply spec/requests/my_request_request_spec.rb]: done

**Review by ViaMarcus — APPROVED** [review]

File: app/models/request.rb (hunk 3 above)
Comment (kermit-klein) [reply app/models/request.rb]: done

File: app/models/request.rb [hunk 6]
```diff
@@ -1,7 +1,8 @@
 # frozen_string_literal: true
 
 class Request < ApplicationRecord
-  validates_presence_of :title, :description, :reward
+  validates_presence_of :title, :description, :reward, :status
   belongs_to :requester, class_name: 'User'
   has_many :offers
+  enum status: { pending: 0, started: 1, active: 2, completed: 3 }
```
Comment (ViaMarcus) [inline app/models/request.rb]: Isn't it pending -> active -> started -> completed?
But it doesn't matter too much, as long as we can keep the unique values

Comment (kermit-klein) [reply app/models/request.rb]: done done

**Review by erikbjoern — APPROVED** [review]

**Pull request merged**